### PR TITLE
[PoC] BLS Aggregate Signature in IBFT ComittedSeal

### PR DESCRIPTION
**Pull request by Kourin1996**
Fix EDGE-550

# Description

This PR adds BLS feature which enables aggregate CommittedSeal signatures.
To use BLS, you will need to start BLS mode like this

```
$ polygon-edge secrets init ... --bls # display BLS Public Key as well
[SECRETS INIT]
Public key (address) = <VALIDATOR_ADDR>
BLS Public key       = <VALIDATOR_BLS_PUBLIC_KEY>
Node ID              = <LIBP2P_NODE ID>

$ polygon-edge genesis --consensus ibft --bls \
--ibft-validator <VALIDATOR_ADDR>:<VALIDATOR_BLS_PUBLIC_KEY> ...

$ polygon-edge server --bls ...
```

With BLS, validator uses 2 signing algorithm ECDSA for Seal and BLS for CommittedSeal because BLS doesn't provide ecrecover, which returns address from the signature.
Currently BLS secrete key is generated from ECDSA private key but each validators should know all validator's BLS public keys to verify CommittedSeal. For that reason, you need to give BLS pubkeys when generating genesis.

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [ ] I have added sufficient documentation in code

## Testing

- [x] I have tested this code with the official test suite
- [x] I have tested this code manually

### Manual tests

Please complete this section if you ran manual tests for this functionality, otherwise delete it

# Documentation update

Please link the documentation update PR in this section if it's present, otherwise delete it

# Additional comments

Please post additional comments in this section if you have them, otherwise delete it

### CODE REVIEW

**Comment by zivkovicmilos** [comment]
@Kourin1996 

Can you merge out the `develop branch` into whatever is the target branch here?

It's hard to follow along changes that are already present on `develop`

File: command/genesis/params.go [hunk 1]
```diff
@@ -197,21 +232,30 @@ func (p *genesisParams) initValidatorSet() error {
 }
 
 func (p *genesisParams) isValidatorNumberValid() bool {
-	return uint64(len(p.ibftValidators)) <= p.maxNumValidators
+	return p.ibftValidators == nil || uint64(p.ibftValidators.Len()) <= p.maxNumValidators
```
Comment (zivkovicmilos) [inline command/genesis/params.go:234]: How is it correct for this check to return ✅ if the `p.ibftValidators` is `nil`?

File: command/genesis/params.go [hunk 2]
```diff
@@ -319,3 +369,69 @@ func (p *genesisParams) getResult() command.CommandResult {
...
+		vals = append(vals, validators.BLSValidator{
+			Address:   types.BytesToAddress(addrBytes),
+			BLSPubKey: pubKeyBytes,
+		})
+	}
+
+	newValSet := validators.BLSValidatorSet(vals)
+
+	return &newValSet, nil
+}
```
Comment (zivkovicmilos) [inline command/genesis/params.go]: What do you think about moving these methods out do some better package, like `validators`?

This is not logic that is specific for the genesis params

File: command/genesis/params.go [hunk 3]
```diff
@@ -319,3 +369,69 @@ func (p *genesisParams) getResult() command.CommandResult {
...
+	}
+
+	for _, v := range values {
+		bytes, err := hex.DecodeString(strings.TrimPrefix(v, "0x"))
+		if err != nil {
+			return nil, err
+		}
+
+		addrs = append(addrs, types.BytesToAddress(bytes))
+	}
```
Comment (zivkovicmilos) [inline command/genesis/params.go]: This method (and the analog `parseBLSValidators`) does way too much:
* It parses a validator set if it's passed in, and uses it as the base validator set
* It parses the validator set in the form of addresses from the params

What do you think about splitting / simplifying this logic up?

cc @dbrajovic

File: command/genesis/utils.go [hunk 4]
```diff
@@ -119,8 +122,47 @@ func getValidatorsFromPrefixPath(prefix string) ([]types.Address, error) {
...
+
+		vals[idx] = validators.BLSValidator{
+			Address:   crypto.PubKeyToAddress(&key.PublicKey),
+			BLSPubKey: pubkey,
+		}
+	}
+
+	valSet := validators.BLSValidatorSet(vals)
+
+	return &valSet, nil
```
Comment (zivkovicmilos) [inline command/genesis/utils.go]: What do you think about moving this out to a more appropriate place?

File: command/secrets/init/params.go [hunk 5]
```diff
@@ -44,10 +50,18 @@ func (ip *initParams) validateFlags() error {
 		return errInvalidParams
 	}
 
+	if ip.rawKeyType == "" {
+		return errInvalidParams
```
Comment (zivkovicmilos) [inline command/secrets/init/params.go]: This error is not appropriate for the key type, as it's related to directory initializations.

Please use a more appropriate error (create a new one) 🙏

File: command/server/server.go [hunk 6]
```diff
@@ -197,6 +198,14 @@ func setFlags(cmd *cobra.Command) {
 		"write all logs to the file at specified location instead of writing them to console",
 	)
 
+	cmd.Flags().BoolVar(
+		&params.useBls,
+		blsFlag,
+		false,
+		// TODO
```
Comment (zivkovicmilos) [inline command/server/server.go]: Leftover TODO

File: server/server.go [hunk 7]
```diff
@@ -405,6 +405,7 @@ func (s *Server) setupConsensus() error {
 			Metrics:        s.serverMetrics.consensus,
 			SecretsManager: s.secretsManager,
 			BlockTime:      s.config.BlockTime,
+			BLS:            s.config.BLS,
```
Comment (zivkovicmilos) [inline server/server.go:409]: Why are we passing it as part of the `server` command configuration?

Can't we use it as part of the consensus `config` in the `genesis.json`?

That way `server` doesn't have to worry about whether BLS is enabled or not

File: secrets/helper/helper.go [hunk 8]
```diff
@@ -79,22 +79,37 @@ func SetupGCPSSM(
...
 
+	address := crypto.PubKeyToAddress(&validatorKey.PublicKey)
+
 	// Write the validator private key to the secrets manager storage
 	if setErr := secretsManager.SetSecret(
 		secrets.ValidatorKey,
 		validatorKeyEncoded,
 	); setErr != nil {
-		return nil, setErr
+		return types.ZeroAddress, setErr
```
Comment (zivkovicmilos) [inline secrets/helper/helper.go:79]: Why `types.ZeroAddress` and not `nil`?

File: secrets/helper/helper.go [hunk 9]
```diff
@@ -79,22 +79,37 @@ func SetupGCPSSM(
...
+
+func GetBLSPubkeyFromValidatorKey(secretsManager secrets.SecretsManager) ([]byte, error) {
+	ecdsaKey, err := crypto.ReadConsensusKey(secretsManager)
+	if err != nil {
+		return nil, err
+	}
+
+	pubkeyBytes, err := crypto.ECDSAToBLSPubkey(ecdsaKey)
+	if err != nil {
+		return nil, err
```
Comment (zivkovicmilos) [inline secrets/helper/helper.go]: Why is this in the `secrets` package, when it only contains crypto functionality?

File: e2e/ibft_test.go [hunk 10]
```diff
@@ -30,6 +32,7 @@ func TestIbft_Transfer(t *testing.T) {
 		func(i int, config *framework.TestServerConfig) {
 			config.Premine(senderAddr, framework.EthToWei(10))
 			config.SetSeal(true)
+			config.SetShowsLog(i == 0)
```
Comment (zivkovicmilos) [inline e2e/ibft_test.go]: This is leftover

File: crypto/crypto.go [hunk 11]
```diff
@@ -299,3 +316,29 @@ func ReadConsensusKey(manager secrets.SecretsManager) (*ecdsa.PrivateKey, error)
...
+
+	_, sk, err := blsPop.KeygenWithSeed(keyBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return sk, nil
+}
+
+func ECDSAToBLSPubkey(ecdsaKey *ecdsa.PrivateKey) ([]byte, error) {
```
Comment (zivkovicmilos) [inline crypto/crypto.go]: Nitpick: Maybe we should call this method something different, as it doesn't really output the Public key, but the marshalled public key.

What do you think about `ECDSAToBLSPublicMarshalled`?

cc @dbrajovic

File: consensus/ibft/ibft.go [hunk 12]
```diff
@@ -32,6 +31,7 @@ var (
 	ErrInvalidHookParam     = errors.New("invalid IBFT hook param passed in")
 	ErrInvalidMechanismType = errors.New("invalid consensus mechanism type in params")
 	ErrMissingMechanismType = errors.New("missing consensus mechanism type in params")
+	ErrSignerNotFound       = errors.New("not found signer in validator set")
```
Comment (zivkovicmilos) [inline consensus/ibft/ibft.go]: Typo here, should be `signer not found in validator set`

File: consensus/ibft/ibft.go [hunk 13]
```diff
@@ -210,12 +222,11 @@ func (i *Ibft) Start() error {
 		proto.RegisterIbftOperatorServer(i.Grpc, i.operator)
 	}
 
-	// Set up the node's validator key
-	if err := i.createKey(); err != nil {
-		return err
-	}
+	i.msgQueue = newMsgQueue()
+	i.closeCh = make(chan struct{})
+	i.updateCh = make(chan struct{})
```
Comment (zivkovicmilos) [inline consensus/ibft/ibft.go]: Just curious, why init it here and now? 👀

Comment (zivkovicmilos) [reply consensus/ibft/ibft.go]: Ah, you moved it from `createKey` which is now deprecated 👍

File: consensus/ibft/ibft.go [hunk 14]
```diff
@@ -998,27 +963,18 @@ func (i *Ibft) updateMetrics(block *types.Block) {
...
-					commit.From,
-				),
-			)
+	committedSeals := make(map[types.Address][]byte)
 
-			continue
+	for addr, commit := range i.state.committed {
+		committedSeal, err := hex.DecodeHex(commit.Seal)
+		if err != nil {
+			return fmt.Errorf("failed to decode committed seal from %s: %w", addr, err)
```
Comment (zivkovicmilos) [inline consensus/ibft/ibft.go]: Do we want this to be fatal?

You dropped the `continue` from this conditional. This means that if you have `1` invalid committed seal, you will go into round change

File: consensus/ibft/ibft.go [hunk 15]
```diff
@@ -1244,9 +1200,9 @@ func (i *Ibft) setState(s IbftState) {
 }
 
 // forceTimeout sets the forceTimeoutCh flag to true
-func (i *Ibft) forceTimeout() {
-	i.forceTimeoutCh = true
-}
+// func (i *Ibft) forceTimeout() {
+// 	i.forceTimeoutCh = true
+// }
```
Comment (zivkovicmilos) [inline consensus/ibft/ibft.go]: Leftover

File: consensus/ibft/ibft.go [hunk 16]
```diff
@@ -1290,7 +1246,14 @@ func (i *Ibft) verifyHeaderImpl(snap *Snapshot, parent, header *types.Header) er
...
+		err = i.signer.VerifyParentCommittedSeal(
+			parentSnap.Set,
+			parent,
+			header,
+			i.quorumSize(header.Number),
+		)
+
+		if err != nil {
 			return fmt.Errorf("failed to verify ParentCommittedSeal: %w", err)
 		}
```
Comment (zivkovicmilos) [inline consensus/ibft/ibft.go]: You can inline this

File: consensus/ibft/ibft.go [hunk 17]
```diff
@@ -1389,6 +1352,18 @@ func (i *Ibft) Close() error {
...
+func (i *Ibft) SetHeaderHash() {
+	types.HeaderHash = func(h *types.Header) types.Hash {
+		hash, err := i.signer.CalculateHeaderHash(h)
+		if err != nil {
+			return types.ZeroHash
+		}
+
+		return hash
+	}
+}
```
Comment (zivkovicmilos) [inline consensus/ibft/ibft.go]: Is there any way we can mitigate setting this **global** callback to something specific?

Usually this is a red flag for bad module design, can we evaluate where we need this `types.HeaderHash`?

cc @dbrajovic

Comment (zivkovicmilos) [reply consensus/ibft/ibft.go]: I know we were setting this global variable previously as well, I'm just wondering if we can go away with this shenanigans 👀

File: consensus/ibft/pos.go [hunk 18]
```diff
@@ -204,13 +206,21 @@ func (pos *PoSMechanism) ShouldWriteTransactions(blockNumber uint64) bool {
...
 	}
 
-	return staking.QueryValidators(transition, pos.ibft.validatorKeyAddr)
+	valAddrs, err := staking.QueryValidators(transition, pos.ibft.signer.Address())
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: PoS contract doesn't hold BLS PubKeys
+	valSet := validators.ECDSAValidatorSet(valAddrs)
```
Comment (zivkovicmilos) [inline consensus/ibft/pos.go]: Leftover TODO?

File: command/server/server.go (hunk 6 above)
Comment (Kourin1996) [reply command/server/server.go]: This is not leftover but we need to update an interface of contract and the its implementation.

File: secrets/helper/helper.go (hunk 8 above)
Comment (Kourin1996) [reply secrets/helper/helper.go:79]: No reason. When this method return error, first return value will be zero value and will not be used. Or do you want caller to check first returning value is nil?

File: crypto/crypto.go (hunk 11 above)
Comment (Kourin1996) [reply crypto/crypto.go]: Not `ECDSAToMarshalledBLSPublic`?

File: consensus/ibft/pos.go (hunk 18 above)
Comment (Kourin1996) [reply consensus/ibft/pos.go]: I commented in other your review

File: consensus/ibft/ibft.go (hunk 17 above)
Comment (Kourin1996) [reply consensus/ibft/ibft.go]: HashCalculation is used widely and I know global variable is not proper

Comment (Kourin1996) [reply consensus/ibft/ibft.go]: We need to figure out which modules are using this

Comment (Kourin1996) [reply consensus/ibft/ibft.go]: But hash calculation completely depends on consensus module due to extra field

File: consensus/ibft/snapshot_test.go [hunk 19]
```diff
@@ -50,36 +52,42 @@ func (t *testerAccount) Address() types.Address {
...
-	h, _ = writeSeal(t.priv, h)
+func (t *testerAccount) sign(h *types.Header) (*types.Header, error) {
+	signer := signer.NewECDSASignerFromKey(t.priv)
 
-	return h
+	return signer.WriteSeal(h)
 }
 
 type testerAccountPool struct {
+	t        *testing.T
```
Comment (zivkovicmilos) [inline consensus/ibft/snapshot_test.go:64]: I honestly don't think we should be dragging a reference to `testing.T` around in the helper methods.

Why not just return errors for methods that can error out?

File: consensus/ibft/signer/signer.go [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+	ErrInvalidValidatorSet        = errors.New("invalid validator set type")
+	ErrInvalidBLSSignature        = errors.New("invalid BLS signature")
+	ErrNilParentHeader            = errors.New("parent header is nil")
+)
+
+// Signer is responsible for signing for blocks and messages in IBFT
+type Signer interface {
+	Address() types.Address
+	InitIBFTExtra(header, parent *types.Header, set validators.ValidatorSet) error
+	GetIBFTExtra(*types.Header) (*IstanbulExtra, error)
```
Comment (zivkovicmilos) [inline consensus/ibft/signer/signer.go:30]: I love the idea of having a separate Signer.

Now we can easily plug and play in tests since the `Signer` is an interface.

Do you think there is any way we can separate out this IBFT specific logic from the signer?
If not right now, just curious what you think on separating it out in the future.

File: consensus/ibft/signer/helper.go [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+	if err != nil {
+		return err
+	}
+
+	msg.From = addr.String()
+
+	return nil
+}
+
+// TODO: move to IBFT?
```
Comment (zivkovicmilos) [inline consensus/ibft/signer/helper.go]: Leftover?

File: consensus/ibft/signer/extra_test.go [hunk 22]
```diff
@@ -0,0 +1,199 @@
+package signer
```
Comment (zivkovicmilos) [inline consensus/ibft/signer/extra_test.go:1]: This entire file is commented out

File: consensus/ibft/signer/ecdsa.go [hunk 23]
```diff
@@ -0,0 +1,362 @@
...
+		return nil, err
+	}
+
+	err = s.packFieldIntoIbftExtra(header, func(ie *IstanbulExtra) {
+		ie.Seal = seal
+	})
+
+	if err != nil {
+		return nil, err
+	}
```
Comment (zivkovicmilos) [inline consensus/ibft/signer/ecdsa.go]: This can be inlined

File: consensus/ibft/signer/ecdsa.go [hunk 24]
```diff
@@ -0,0 +1,362 @@
...
+		return ErrInvalidCommittedSealType
+	}
+
+	validatorSet, ok := rawSet.(*validators.ECDSAValidatorSet)
+	if !ok {
+		return ErrInvalidValidatorSet
+	}
+
+	// get the message that needs to be signed
+	// this not signing! just removing the fields that should be signed
```
Comment (zivkovicmilos) [inline consensus/ibft/signer/ecdsa.go]: Deprecated comment?

File: consensus/ibft/signer/ecdsa.go [hunk 25]
```diff
@@ -0,0 +1,362 @@
...
+		return ErrInvalidCommittedSealType
+	}
+
+	validatorSet, ok := rawParentSet.(*validators.ECDSAValidatorSet)
+	if !ok {
+		return ErrInvalidValidatorSet
+	}
+
+	// get the message that needs to be signed
+	// this not signing! just removing the fields that should be signed
```
Comment (zivkovicmilos) [inline consensus/ibft/signer/ecdsa.go]: Deprecated comment?

File: consensus/ibft/signer/ecdsa.go [hunk 26]
```diff
@@ -0,0 +1,362 @@
...
+
+	putIbftExtra(h, extra)
+
+	return nil
+}
+
+func (s *ECDSASigner) filterHeaderForHash(header *types.Header) (*types.Header, error) {
+	// This will effectively remove the Seal and Committed Seal fields,
+	// while keeping proposer vanity and validator set
+	// because extra.Validators, extra.ParentCommittedSeal is what we got from `h` in the first place.
```
Comment (zivkovicmilos) [inline consensus/ibft/signer/ecdsa.go]: Deprecated comment

File: consensus/ibft/signer/ecdsa.go [hunk 27]
```diff
@@ -0,0 +1,362 @@
...
+		if !validators.Includes(addr) {
+			return ErrNonValidatorCommittedSeal
+		}
+
+		visited[addr] = struct{}{}
+	}
+
+	// Valid committed seals must be at least 2F+1
+	// 	2F 	is the required number of honest validators who provided the committed seals
+	// 	+1	is the proposer
```
Comment (zivkovicmilos) [inline consensus/ibft/signer/ecdsa.go]: This comment is no longer valid

cc @dbrajovic

File: consensus/ibft/signer/bls.go [hunk 28]
```diff
@@ -0,0 +1,462 @@
...
+}
+
+func (s *BLSSigner) Address() types.Address {
+	return s.address
+}
+
+func (s *BLSSigner) InitIBFTExtra(header, parent *types.Header, set validators.ValidatorSet) error {
+	var parentCommittedSeal Sealer
+
+	if header.Number > 1 {
```
Comment (zivkovicmilos) [inline consensus/ibft/signer/bls.go]: Why is this check even required?

File: consensus/ibft/signer/bls.go [hunk 29]
```diff
@@ -0,0 +1,462 @@
...
+		return ErrInvalidCommittedSealType
+	}
+
+	validatorSet, ok := rawSet.(*validators.BLSValidatorSet)
+	if !ok {
+		return ErrInvalidValidatorSet
+	}
+
+	// get the message that needs to be signed
+	// this not signing! just removing the fields that should be signed
```
Comment (zivkovicmilos) [inline consensus/ibft/signer/bls.go]: Deprecated?

Same as with `ecdsa.go`, for the remaining comments

File: consensus/ibft/signer/bls.go [hunk 30]
```diff
@@ -0,0 +1,462 @@
...
+
+// 	sigBytes, err := sig.MarshalBinary()
+// 	if err != nil {
+// 		return err
+// 	}
+
+// 	msg.Signature = hex.EncodeToHex(sigBytes)
+
+// 	return nil
+// }
```
Comment (zivkovicmilos) [inline consensus/ibft/signer/bls.go]: Leftover?

File: consensus/ibft/validators/bls.go [hunk 31]
```diff
@@ -0,0 +1,233 @@
...
+
+// Add adds a new address to the validator set
+func (v *BLSValidatorSet) Add(addr types.Address) {
+	// TODO: doesn't support right now
+}
+
+// Del removes an address from the validator set
+func (v *BLSValidatorSet) Del(target types.Address) {
+	// TODO: doesn't support right now
+}
```
Comment (zivkovicmilos) [inline consensus/ibft/validators/bls.go]: Does this mean that the set can't ever be changed from the start?

File: consensus/ibft/signer/bls.go (hunk 28 above)
Comment (Kourin1996) [reply consensus/ibft/signer/bls.go]: Because genesis block doesn't have CommittedSeal

File: consensus/ibft/validators/bls.go (hunk 31 above)
Comment (Kourin1996) [reply consensus/ibft/validators/bls.go]: Do you agree with changing type of Miner in Header from Address to []byte?
@zivkovicmilos

**Comment by Kourin1996** [comment]
Clean up Signer in https://github.com/0xPolygon/polygon-edge/pull/574/commits/b2a4844949c1d111d6bc3ea959291e44fc47cdcb

ECDSASigner and BLSSigner have redundant codes, so I renamed them to ECDSAKeyManager, BLSManager and added new Signer. Now Signer defines the flow and ECDSASigner and BLSSigner defines the implementation

File: command/genesis/params.go (hunk 2 above)
Comment (Kourin1996) [reply command/genesis/params.go]: Good idea. Moved at b8b41249444ab5cda3c156f251c6a9ced84bb5f1

File: command/genesis/params.go (hunk 3 above)
Comment (Kourin1996) [reply command/genesis/params.go]: In b8b41249444ab5cda3c156f251c6a9ced84bb5f1, I remove appending from this function

File: command/genesis/utils.go (hunk 4 above)
Comment (Kourin1996) [reply command/genesis/utils.go]: Good catch. Moved in 54c8009a3f1e8d324502565a9aff5bf57d3697b0

File: command/secrets/init/params.go (hunk 5 above)
Comment (Kourin1996) [reply command/secrets/init/params.go]: Added new error in 5c1ee77068408fb15fdddbcf3a0e77d14f46abc8

File: secrets/helper/helper.go (hunk 9 above)
Comment (Kourin1996) [reply secrets/helper/helper.go]: Moved in b12f4bdc7253b1bbb514a566e2e1bb25b2602b1d

File: e2e/ibft_test.go (hunk 10 above)
Comment (Kourin1996) [reply e2e/ibft_test.go]: Removed in 1cb50b199da5a3e25ef7293593079cef6bf454b5

File: crypto/crypto.go (hunk 11 above)
Comment (Kourin1996) [reply crypto/crypto.go]: Renamed in ab0ac775cd9fc1db493b913b2cf8d0239da4638a

File: consensus/ibft/ibft.go (hunk 12 above)
Comment (Kourin1996) [reply consensus/ibft/ibft.go]: Fixed in b5b0944399257ea29d9b1210cc16dcab4344aeea

File: consensus/ibft/ibft.go (hunk 14 above)
Comment (Kourin1996) [reply consensus/ibft/ibft.go]: Followed latest code in d9af3b51e19bf7d8ab606f4a549fbaaf68a19113

File: consensus/ibft/ibft.go (hunk 15 above)
Comment (Kourin1996) [reply consensus/ibft/ibft.go]: Fixed in a90c04b73f85d87689c1701e24cc87b156a06331

File: consensus/ibft/ibft.go (hunk 16 above)
Comment (Kourin1996) [reply consensus/ibft/ibft.go]: Fixed in c04ef82713935e921b1f320d7788a70dde7cf307

File: consensus/ibft/signer/helper.go (hunk 21 above)
Comment (Kourin1996) [reply consensus/ibft/signer/helper.go]: Removed in c04ef82713935e921b1f320d7788a70dde7cf307

File: consensus/ibft/signer/ecdsa.go (hunk 23 above)
Comment (Kourin1996) [reply consensus/ibft/signer/ecdsa.go]: Fixed in c04ef82713935e921b1f320d7788a70dde7cf307

File: consensus/ibft/signer/ecdsa.go (hunk 24 above)
Comment (Kourin1996) [reply consensus/ibft/signer/ecdsa.go]: Removed

File: consensus/ibft/signer/ecdsa.go (hunk 25 above)
Comment (Kourin1996) [reply consensus/ibft/signer/ecdsa.go]: Removed

File: consensus/ibft/signer/ecdsa.go (hunk 26 above)
Comment (Kourin1996) [reply consensus/ibft/signer/ecdsa.go]: No, but I moved below c04ef82713935e921b1f320d7788a70dde7cf307

File: consensus/ibft/signer/bls.go (hunk 29 above)
Comment (Kourin1996) [reply consensus/ibft/signer/bls.go]: Removed

File: consensus/ibft/signer/bls.go (hunk 30 above)
Comment (Kourin1996) [reply consensus/ibft/signer/bls.go]: Removed

File: consensus/ibft/signer/extra_test.go (hunk 22 above)
Comment (Kourin1996) [reply consensus/ibft/signer/extra_test.go:1]: Uncommented in 4ba1feaf1bdaa053ece5b8a4c3ffd7ecd4a19362

**Comment by Kourin1996** [comment]
@zivkovicmilos Please resolve comments

File: secrets/helper/helper.go (hunk 8 above)
Comment (dbrajovic) [reply secrets/helper/helper.go:79]: Also, you can't use `nil` when the return type is `types.Address`

File: consensus/ibft/snapshot.go [hunk 32]
```diff
@@ -254,7 +256,72 @@ type Snapshot struct {
...
+	if err := json.Unmarshal(data, &raw); err != nil {
+		return err
+	}
+
+	s.Number = raw.Number
+	s.Hash = raw.Hash
+	s.Votes = raw.Votes
+
+	if len(raw.Set) > 0 {
+		switch raw.Set[0].(type) {
```
Comment (dbrajovic) [inline consensus/ibft/snapshot.go:350]: Maybe we could extract this switch logic into a separate helper ?

File: consensus/ibft/ibft.go [hunk 33]
```diff
@@ -998,12 +954,11 @@
...
 
 func (i *Ibft) insertBlock(block *types.Block) error {
-	committedSeals := make([][]byte, 0)
+	committedSeals := make(map[types.Address][]byte)
 
-	for _, commit := range i.state.committed {
-		// no need to check the format of seal here because writeCommittedSeals will check
-		committedSeal, decodeErr := hex.DecodeHex(commit.Seal)
-		if decodeErr != nil {
+	for addr, commit := range i.state.committed {
```
Comment (dbrajovic) [inline consensus/ibft/ibft.go]: Readability nitpick: make this a separate method (extractSeals)

File: consensus/ibft/ibft.go (hunk 17 above)
Comment (dbrajovic) [reply consensus/ibft/ibft.go]: Given how widespread usage of `Header`'s `ComputeHash()` method is, I'd say this is best we can do right now. 

A good direction to think about this is how many modules would actually need the hash of a given `Header`. If that list isn't too long, then maybe this implementation could be more specific.

File: consensus/ibft/validators/bls.go [hunk 34]
```diff
@@ -0,0 +1,303 @@
...
+	return val
+}
+
+func (v *BLSValidator) UnmarshalRLPFrom(p *fastrlp.Parser, vv *fastrlp.Value) error {
+	vals, err := vv.GetElems()
+	if err != nil {
+		return fmt.Errorf("mismatch of RLP type for CommittedSeal, expected list but found %s", vv.Type())
+	}
+
+	{
```
Comment (dbrajovic) [inline consensus/ibft/validators/bls.go]: Nitpick: maybe the wrapping scope braces can be removed ?

File: consensus/ibft/validators/bls.go [hunk 35]
```diff
@@ -0,0 +1,303 @@
...
+func (v *BLSValidatorSet) UnmarshalRLPFrom(p *fastrlp.Parser, vv *fastrlp.Value) error {
+	vals, err := vv.GetElems()
+	if err != nil {
+		return fmt.Errorf("mismatch of RLP type for ECDSAValidatorSet, expected list but found %s", vv.Type())
+	}
+
+	valSet := make([]BLSValidator, len(vals))
+
+	for index := range vals {
+		vals := vals
```
Comment (dbrajovic) [inline consensus/ibft/validators/bls.go]: Why the identity assignment ?

File: consensus/ibft/signer/bls.go (hunk 28 above)
Comment (zivkovicmilos) [reply consensus/ibft/signer/bls.go]: Of course 🤦‍♂️

File: consensus/ibft/validators/bls.go (hunk 31 above)
Comment (zivkovicmilos) [reply consensus/ibft/validators/bls.go]: Honestly, I'm completely fine with whatever gives us greater flexibility. If it's `[]byte` than I'm cool with that 🙂 

cc @dbrajovic @Aleksao998

File: consensus/ibft/signer/ecdsa.go [hunk 36]
```diff
@@ -0,0 +1,182 @@
...
+	committedSeal *SerializedSeal,
+	msg []byte,
+	validators validators.ECDSAValidatorSet,
+) (int, error) {
+	numSeals := len(*committedSeal)
+	if numSeals == 0 {
+		return 0, ErrEmptyCommittedSeals
+	}
+
+	visited := map[types.Address]struct{}{}
```
Comment (dbrajovic) [inline consensus/ibft/signer/ecdsa.go]: What about a map of booleans (set) ?

File: consensus/ibft/pos.go [hunk 37]
```diff
@@ -204,13 +206,21 @@ func (pos *PoSMechanism) ShouldWriteTransactions(blockNumber uint64) bool {
...
 		return nil, err
 	}
 
-	return staking.QueryValidators(transition, pos.ibft.validatorKeyAddr)
+	valAddrs, err := staking.QueryValidators(transition, pos.ibft.signer.Address())
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: PoS contract doesn't hold BLS PubKeys
```
Comment (dbrajovic) [inline consensus/ibft/pos.go]: Is this a reference for a future feature ?

**Review by dbrajovic — CHANGES_REQUESTED** [review]
Good job on encapsulating the `signer` logic from the rest of ibft. On that note, what do you think about the following:

Since the `signer` package defines an interface which it implements with `KeyManager` under the hood (`SignerImpl`), maybe this is a good opportunity for custom constructors ? (e.g. `NewBLSSigner`, `NewECDSASigner`). The constructors would still return a `Signer` type, preserving the used api. 

This would add great flexibility should there be a need for another implementation. (I realize it might be a hassle to change this now, just interested in your opinion).

cc @zivkovicmilos

**Comment by Kourin1996** [comment]
@dbrajovic I'm not sure. Custom contract is different feature. At least signer will not depend on its implementation. Signer should focus on only signing and validators will be managed by different module

**Comment by dbrajovic** [comment]
> At least signer will not depend on its implementation.


@Kourin1996 
Exactly. `ibft` package sees just the `Signer` interface, in which case the `InitSigner` can be simplified to:

```
func InitSigner(secretManager secrets.SecretsManager, isBLS bool) (signer.Signer, error) {
	if isBLS {
		return NewBLSSigner(secretManager)
	}
	
	return NewECDSASigner(secretManager)
}
````

**Comment by lazartravica** [comment]
@Kourin1996 we can move this to draft until the backwards compatibility is clear.

**Comment by zivkovicmilos** [comment]
@Kourin1996 

Please pull in the latest develop and resolve any conflicts 🙏

File: server/server.go (hunk 7 above)
Comment (Kourin1996) [reply server/server.go:409]: Resolving because this code has changed in the next PR https://github.com/0xPolygon/polygon-edge/pull/649.
Basically IBFT Hook module takes this configurations

File: consensus/ibft/validators/bls.go (hunk 31 above)
Comment (Kourin1996) [reply consensus/ibft/validators/bls.go]: Resolving because this stuff is done in this PR https://github.com/0xPolygon/polygon-edge/pull/649

File: consensus/ibft/snapshot.go (hunk 32 above)
Comment (Kourin1996) [reply consensus/ibft/snapshot.go:350]: Resolving this because this issue is resolved in the next PR https://github.com/0xPolygon/polygon-edge/pull/649

File: consensus/ibft/pos.go (hunk 37 above)
Comment (Kourin1996) [reply consensus/ibft/pos.go]: Resolving because this issue is fixed in the next PR https://github.com/0xPolygon/polygon-edge/pull/649

File: consensus/ibft/signer/ecdsa.go (hunk 36 above)
Comment (Kourin1996) [reply consensus/ibft/signer/ecdsa.go]: Good idea. Fixed at 52379b35c037308d4ff8e5b77253cacc292214f1

File: consensus/ibft/validators/bls.go (hunk 35 above)
Comment (Kourin1996) [reply consensus/ibft/validators/bls.go]: Resolving because this code has been removed already

File: consensus/ibft/validators/bls.go (hunk 34 above)
Comment (Kourin1996) [reply consensus/ibft/validators/bls.go]: Resolving because this code has changed already

File: validators/types.go [hunk 38]
```diff
@@ -0,0 +1,96 @@
...
+	case ECDSAValidatorType, BLSValidatorType:
+		*t = x
+
+		return nil
+	default:
+		return fmt.Errorf("invalid validator type: %s", s)
+	}
+}
+
+func NewValidatorFromType(t ValidatorType) Validator {
```
Comment (dbrajovic) [inline validators/types.go:38]: Unused ctor

File: validators/types.go [hunk 39]
```diff
@@ -0,0 +1,96 @@
...
+
+	switch x {
+	case ECDSAValidatorType, BLSValidatorType:
+		*t = x
+
+		return nil
+	default:
+		return fmt.Errorf("invalid validator type: %s", s)
+	}
+}
```
Comment (dbrajovic) [inline validators/types.go:36]: Nitpick: replace `FromString` method with `ValidatorTypeFromString` free func.

```
func ValidatorTypeFromString(s string) (ValidatorType, error) {
	vt := ValidatorType(s)

	switch vt {
	case ECDSAValidatorType, BLSValidatorType:
		return vt, nil
	default:
		return "", errors.New("asd")
	}
}
```

File: types/types.go [hunk 40]
```diff
@@ -150,6 +150,10 @@ func BytesToAddress(b []byte) Address {
 	return a
 }
 
+func AddressToBytes(a Address) []byte {
```
Comment (dbrajovic) [inline types/types.go:153]: Unused

File: validators/types.go (hunk 38 above)
Comment (Kourin1996) [reply validators/types.go:38]: Removed in the next PR, a85e43ec5b5186573f08192d7be09d99242748b4

File: validators/types.go (hunk 39 above)
Comment (Kourin1996) [reply validators/types.go:36]: Can you tell me what the difference is?

File: types/types.go (hunk 40 above)
Comment (Kourin1996) [reply types/types.go:153]: Removed in the next PR, a42d46541be2da9a1ad487aec790e151ecd06fad
